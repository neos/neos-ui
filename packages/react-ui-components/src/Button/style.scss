@import '../reset';

.btn {
    @include reset;
    position: relative;
    display: inline-block;
    height: var(--spacing-GoldenUnit);
    min-width: var(--spacing-GoldenUnit);
    padding: 0 var(--spacing-Full);
    border: 0;
    -webkit-font-smoothing: antialiased;
    vertical-align: middle;
    width: auto;
    color: var(--colors-ContrastBrightest);
    cursor: pointer;
    background-color: var(--colors-ContrastNeutral);
    font-size: var(--fontSize-Base);

    &:focus {
        outline: 1px solid var(--colors-PrimaryBlue);
    }

    &[disabled] {
        opacity: .5;
        cursor: not-allowed;

        &:hover,
        &.btn--isPressed,
        &:hover i,
        i {
            color: var(--colors-ContrastBrightest);
        }
    }
}
.btn--size-small {
    height: 32px;
    font-size: 13px;
    padding: 0 8px;
}

.btn--clean {
    background-color: var(--colors-ContrastDarker);
}
.btn--lighter,
.btn--lighter i {
    background-color: var(--colors-ContrastDark);
}
.btn--neutral,
.btn--neutral i {
    background-color: var(--colors-ContrastNeutral);
}
.btn--transparent {
    background: 0;
}
.btn--brand,
.btn--brand i {
    background-color: var(--colors-PrimaryBlue);
    color: white !important;
}
.btn--success,
.btn--success i {
    background-color: var(--colors-Success);
    color: white !important;
}
.btn--warn,
.btn--warn i {
    background-color: var(--colors-Warn);
}
.btn--error,
.btn--error i {
    background-color: var(--colors-Error);
    color: white !important;
}

.btn--brandActive {
    background-color: var(--colors-PrimaryBlue);

    svg {
        color: var(--colors-ContrastBrightest);
    }
}

.btn--brandHover:hover:enabled,
.btn--brandHover:hover:enabled i,
.btn--brandHover.btn--isPressed,
.btn--brandHover.btn--isPressed i {
    background-color: var(--colors-PrimaryBlue);
}
.btn--cleanHover.btn--isPressed,
.btn--cleanHover.btn--isPressed i,
.btn--cleanHover:hover:enabled,
.btn--cleanHover:hover:enabled i {
    color: var(--colors-PrimaryBlue);

    svg {
        color: var(--colors-PrimaryBlue);
    }
}
.btn--darkenHover.btn--isPressed,
.btn--darkenHover.btn--isPressed i,
.btn--darkenHover:hover:enabled,
.btn--darkenHover:hover:enabled i {
    background-color: rgba(0, 0, 0, .1);
}
.btn--successHover.btn--isPressed,
.btn--successHover.btn--isPressed i,
.btn--successHover:hover:enabled,
.btn--successHover:hover:enabled i {
    background-color: var(--colors-SuccessHover);
}
.btn--warnHover.btn--isPressed,
.btn--warnHover.btn--isPressed i,
.btn--warnHover:hover:enabled,
.btn--warnHover:hover:enabled i {
    background-color: var(--colors-WarnHover);
}
.btn--errorHover:hover:enabled,
.btn--errorHover:hover:enabled i,
.btn--errorHover.btn--isPressed,
.btn--errorHover.btn--isPressed i {
    background-color: var(--colors-ErrorHover);
}
