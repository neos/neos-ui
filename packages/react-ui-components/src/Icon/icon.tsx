import React, {PureComponent} from 'react';

import {FontAwesomeIconProps} from '@fortawesome/react-fontawesome';
import {PickDefaultProps, Omit} from '../../types';
import FontAwesomeIcon from './fontAwesomeIcon';
import ResourceIcon, {ResourceIconProps} from './resourceIcon';

type IconSize = 'xs' | 'sm' | 'lg' | '2x' | '3x';
type IconPadding = 'none' | 'left' | 'right';
type IconColor = 'default' | 'warn' | 'error' | 'primaryBlue';

export interface IconTheme {
    readonly icon: string;
    readonly 'icon--huge': string;
    readonly 'icon--large': string;
    readonly 'icon--big': string;
    readonly 'icon--small': string;
    readonly 'icon--tiny': string;
    readonly 'icon--paddedLeft': string;
    readonly 'icon--paddedRight': string;
    readonly 'icon--spin': string;
    readonly 'icon--color-warn': string;
    readonly 'icon--color-error': string;
    readonly 'icon--color-primaryBlue': string;
}

export interface IconProps extends Omit<FontAwesomeIconProps, 'icon'> {
    /**
     * We use the react component FortAwesome provides to render icons.
     * we will pass down all props to the component via {...rest} to expose it's api
     * https://github.com/FortAwesome/react-fontawesome/blob/master/README.md
     */

    /**
     * The identifier of the icon to render.
     * Can be a font-awesome icon identifier or a asset resource.
     */
    readonly icon?: string;

    /**
     * The (accessibility) label for this icon
     */
    readonly label?: string;

    /**
     * Controls the padding around the icon in a standardized way.
     */
    readonly padded?: IconPadding;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     *  Adjust the color of the icon
     */
    readonly color?: IconColor;

    /**
     *  Adjust the size of the icon
     */
    readonly size?: IconSize;

    /**
     *  An optional css theme to be injected.
     */
    readonly theme?: IconTheme;
}

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded' | 'size'>;

export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none',
    size: 'sm'
};

class Icon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {icon} = this.props;

        if (icon && icon.substr(0, 11) === 'resource://') {
            return <ResourceIcon {...this.props as ResourceIconProps} />;
        } else {
            return <FontAwesomeIcon {...this.props} />;
        }
    }
}

export default Icon;
