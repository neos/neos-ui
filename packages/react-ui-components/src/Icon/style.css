.icon {
    composes: reset from './../reset.css';
    display: inline-block;
    font: normal normal normal FontAwesome;
    font-size: 14px/1;
    text-rendering: auto;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
}
.icon--paddedLeft {
    margin-left: .75em;
}
.icon--paddedRight {
    margin-right: .75em;
}

.icon--color-warn {
    color: var(--colors-Warn);
}

.icon--color-error {
    color: var(--colors-Error);
}

.icon--color-primaryBlue {
    color: var(--colors-PrimaryBlue);
}

.icon--huge {
    svg {
        height: 3em;
    }
}

.icon--large {
    svg {
        height: 2em;
    }
}

.icon--big {
    svg {
        height: 1.33em;
    }
}

.icon--small {
    svg {
        height: 1em;
        max-height: 1em;
        max-width: 100%;
    }
}

.icon--tiny {
    svg {
        height: .75em;
    }
}

.icon--resource {
    display: inline-grid;
    width: 100%;
    justify-content: center;
}
