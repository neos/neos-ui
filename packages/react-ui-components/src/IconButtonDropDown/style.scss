@import '../reset';

.wrapper {
    @include reset;
    position: relative;
    display: inline-block;
}
.wrapper__btn {
    @include reset;
    position: relative;
    width: var(--spacing-GoldenUnit);
}
.wrapper__btnModeIcon {
    @include reset;
    position: absolute;
    right: 4px;
    bottom: 4px;
    color: var(--colors-PrimaryBlue);

    [disabled] > & {
        opacity: 0;
    }
}
.wrapper:hover .wrapper__btnIcon {
    color: var(--colors-PrimaryBlue);
}
.wrapper__dropDown {
    @include reset;
    position: absolute;
    top: 100%;
    left: 0;
    z-index: var(--zIndex-WrapperDropdown-Context);
    display: none;
    background: var(--colors-ContrastDarker);
    border: 1px solid var(--colors-ContrastDark);
    border-top: 0;
}
.wrapper__dropDown--isOpen {
    display: block;
}
.wrapper__dropDownItem {
    @include reset;
    position: relative;
    display: block;
    width: calc(var(--spacing-GoldenUnit) - 2px);
    height: calc(var(--spacing-GoldenUnit) - 2px);
    text-align: center;
    vertical-align: middle;
    cursor: pointer;

    &:hover,
    &:hover > * {
        color: var(--colors-PrimaryBlue);
    }

    > * {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translateX(-50%) translateY(-50%);
    }
}
