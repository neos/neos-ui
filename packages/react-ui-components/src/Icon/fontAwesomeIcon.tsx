import {config, IconName, IconPrefix, library} from '@fortawesome/fontawesome-svg-core';
import '@fortawesome/fontawesome-svg-core/styles.css';
import {fab} from '@fortawesome/free-brands-svg-icons';
import {far} from '@fortawesome/free-regular-svg-icons';
import {fas} from '@fortawesome/free-solid-svg-icons';

import {FontAwesomeIcon as FontAwesomeIconOriginComponent} from '@fortawesome/react-fontawesome';
import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';

import {IconProps, defaultProps} from './icon';
import mapper from './mapper';

// tslint:disable:no-object-mutation
config.autoAddCss = false;
config.familyPrefix = 'neos-fa' as IconPrefix;
config.replacementClass = 'neos-svg-inline--fa';
// tslint:enable:no-object-mutation

library.add(fab, fas, far);

class FontAwesomeIcon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {padded, theme, label, icon, className, color, ...rest} = this.props;
        const iconClassName = icon;
        const classNames = mergeClassNames(
            theme!.icon,
            iconClassName,
            className,
            {
                [theme!['icon--paddedLeft']]: padded === 'left',
                [theme!['icon--paddedRight']]: padded === 'right',
                [theme!['icon--color-warn']]: color === 'warn',
                [theme!['icon--color-error']]: color === 'error',
                [theme!['icon--color-primaryBlue']]: color === 'primaryBlue',
            }
        );

        return <FontAwesomeIconOriginComponent icon={icon ? this.getIconProp(icon) as any : 'question'} aria-label={label} className={classNames} {...rest} />;
    }

    private readonly getIconProp = (icon: string): any => {
        const mappedIcon = mapper(icon);
        const iconArray = mappedIcon.split(' ');
        if (iconArray.length > 1) {
            const prefix = iconArray[0];
            const processedIcon = iconArray[1].startsWith('fa-') ? iconArray[1].substr(3) : iconArray[1];
            return [prefix as IconPrefix, processedIcon as IconName];
        } else {
            const prefix: IconPrefix = 'fas';
            return [prefix, mappedIcon as IconName];
        }
    }
}

export default FontAwesomeIcon;
