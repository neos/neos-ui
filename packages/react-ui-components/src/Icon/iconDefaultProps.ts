import {IconProps} from './icon';
import {PickDefaultProps} from '../../types';

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded' | 'size'>;

// extracted because of resulting in a circular dependency in jest
// internal - use `import {defaultProps} from "./icon"` instead
export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none',
    size: 'sm'
};
