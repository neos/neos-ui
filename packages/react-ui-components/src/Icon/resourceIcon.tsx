import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';
import { ReactSVG } from 'react-svg';

import {IconProps, IconTheme, defaultProps} from './icon';
import {Omit} from '../../types';

interface ResourceIconTheme extends IconTheme {
    readonly 'icon--resource': string;
}

export interface ResourceIconProps extends Omit<IconProps, 'theme'> {
    readonly theme?: ResourceIconTheme;
}

class ResourceIcon extends PureComponent<ResourceIconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {padded, theme, label, icon, className, color, size} = this.props;

        if (!icon || icon.substr(0, 11) !== 'resource://') {
            return null;
        }

        const iconResourcePath = '/_Resources/Static/Packages/' + icon.substr(11);
        const classNames = mergeClassNames(
            theme!.icon,
            className,
            {
                [theme!['icon--resource']]: icon,
                [theme!['icon--paddedLeft']]: padded === 'left',
                [theme!['icon--paddedRight']]: padded === 'right',
                [theme!['icon--color-warn']]: color === 'warn',
                [theme!['icon--color-error']]: color === 'error',
                [theme!['icon--color-primaryBlue']]: color === 'primaryBlue',
                [theme!['icon--huge']]: size === '3x',
                [theme!['icon--large']]: size === '2x',
                [theme!['icon--big']]: size === 'lg',
                [theme!['icon--small']]: size === 'sm',
                [theme!['icon--tiny']]: size === 'xs'
            }
        );

        return <ReactSVG src={iconResourcePath} aria-label={label} className={classNames} wrapper="span" />;
    }
}

export default ResourceIcon;
