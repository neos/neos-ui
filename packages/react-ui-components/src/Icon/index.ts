import {themr} from '@friendsofreactjs/react-css-themr';

import identifiers from '../identifiers';
import Icon from './icon';

import style from './style.css';

export default themr(identifiers.icon, style)(Icon);
