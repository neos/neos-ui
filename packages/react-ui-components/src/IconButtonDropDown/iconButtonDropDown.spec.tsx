import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import Button from '../Button';
import Icon from '../Icon';
import DropDownItem from './dropDownItem';
import IconButtonDropDown, {defaultProps, IconButtonDropDownProps} from './iconButtonDropDown';

describe('<IconButtonDropDown/>', () => {
    const props: IconButtonDropDownProps = {
        ...defaultProps,
        children: [
            // @ts-ignore
            // tslint:disable-next-line:jsx-key
            <div dropDownId="foo1"/>,
            // @ts-ignore
            // tslint:disable-next-line:jsx-key
            <div dropDownId="foo2"/>
        ],
        icon: 'barGeneralIcon',
        modeIcon: 'fooModeIcon',
        onClick: jest.fn(),
        onItemSelect: jest.fn(),
        theme: {
            'wrapper': 'baseWrapperClassName',
            'wrapper__btn': 'baseBtnClassName',
            'wrapper__btnIcon': 'baseBtnIconClassName',
            'wrapper__btnModeIcon': 'baseBtnModeIconClassName',
            'wrapper__dropDown': 'baseDropDownClassName',
            'wrapper__dropDown--isOpen': 'baseDropDownOpenClassName',
            'wrapper__dropDownItem': 'baseIconButtonDropDownClassName'
        },
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<IconButtonDropDown {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should initialize with a state of "{isOpen: false}".', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);

        expect(wrapper.state('isOpen')).toBe(false);
    });

    it('should render a "ButtonComponent" component with style="clean" and aria-haspopup="true" prop and a className which matches the themes "wrapper__btn".', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const btn = wrapper.find(Button);

        expect(btn.prop('style')).toBe('clean');
        expect(btn.prop('aria-haspopup')).toBe('true');
        expect(btn.hasClass('baseBtnClassName')).toBeTruthy();
    });

    it('should render a "ButtonComponent" component which reflects the "disabled" prop.', () => {
        const wrapper = shallow(<IconButtonDropDown {...props} disabled={false}/>);
        const btn = wrapper.find(Button);

        expect(btn.prop('disabled')).toBe(false);
    });

    it('should set the "isOpen" state to "true" when pressing the "ButtonComponent" for more than 200 ms.', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const btn = wrapper.find(Button);

        return new Promise<void>(resolve => {
            btn.simulate('mouseDown');

            setTimeout(() => {
                expect(wrapper.state('isOpen')).toBe(true);
                resolve();
            }, 300);
        });
    });

    it('should abort the setting of the "isOpen" state to "true" when pressing and afterwards clicking on the "ButtonComponent" within the 200 ms.', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const btn = wrapper.find(Button);

        return new Promise<void>(resolve => {
            btn.simulate('mouseDown');

            btn.simulate('click');

            setTimeout(() => {
                expect(wrapper.state('isOpen')).toBe(false);
                resolve();
            }, 300);
        });
    });

    it('should call the "onClick" prop when clicking on the "ButtonComponent".', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<IconButtonDropDown {...props} onClick={onClick}/>);
        const btn = wrapper.find(Button);

        btn.simulate('click');

        expect(onClick.mock.calls.length).toBe(1);
    });

    it('should render two "IconComponent"s within the "ButtonComponent".', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const btn = wrapper.find(Button);
        const icons = btn.find(Icon);

        expect(icons.length).toBe(2);
    });
    it('should propagate the "modeIcon" prop to the first "IconComponent".', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const btn = wrapper.find(Button);
        const icon = btn.find(Icon).at(0);

        expect(icon.prop('icon')).toBe('fooModeIcon');
    });
    it('should propagate the "icon" prop to the second "IconComponent".', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const btn = wrapper.find(Button);
        const icon = btn.find(Icon).at(1);

        expect(icon.prop('icon')).toBe('barGeneralIcon');
    });

    it('should render a "DropDownItem" for each passed child and propagate the "dropDownId" to it as "id".', () => {
        const wrapper = shallow(<IconButtonDropDown {...props}/>);
        const items = wrapper.find(DropDownItem);

        expect(items.length).toBe(2);
        expect(items.at(0).prop('id')).toBe('foo1');
        expect(items.at(1).prop('id')).toBe('foo2');
    });
    it('should call the "onItemSelect" prop when clicking on a "DropDownItem".', () => {
        const onItemSelect = jest.fn();
        const wrapper = shallow(<IconButtonDropDown {...props} onItemSelect={onItemSelect}/>);
        const item = wrapper.find(DropDownItem).at(0);

        item.simulate('click');

        expect(onItemSelect.mock.calls.length).toBe(1);
    });
});
