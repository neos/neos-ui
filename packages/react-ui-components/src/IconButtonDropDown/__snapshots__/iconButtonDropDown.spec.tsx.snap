// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButtonDropDown/> should render correctly. 1`] = `
<div
  className="baseWrapperClassName"
  onMouseLeave={[Function]}
>
  <ThemedButton
    _refHandler={[Function]}
    aria-haspopup="true"
    className="baseBtnClassName"
    composeTheme="deeply"
    disabled={false}
    hoverStyle="clean"
    isActive={false}
    isFocused={false}
    mapThemrProps={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    size="regular"
    style="clean"
    type="button"
  >
    <ThemedIcon
      className="baseBtnModeIconClassName"
      color="default"
      composeTheme="deeply"
      icon="fooModeIcon"
      mapThemrProps={[Function]}
      padded="none"
      size="sm"
    />
    <ThemedIcon
      className="baseBtnIconClassName"
      color="default"
      composeTheme="deeply"
      icon="barGeneralIcon"
      mapThemrProps={[Function]}
      padded="none"
      size="sm"
    />
  </ThemedButton>
  <div
    aria-hidden="true"
    className="baseDropDownClassName"
  >
    <DropDownItem
      className="baseIconButtonDropDownClassName"
      id="foo1"
      key="0"
      onClick={[Function]}
    >
      <div
        dropDownId="foo1"
      />
    </DropDownItem>
    <DropDownItem
      className="baseIconButtonDropDownClassName"
      id="foo2"
      key="1"
      onClick={[Function]}
    >
      <div
        dropDownId="foo2"
      />
    </DropDownItem>
  </div>
</div>
`;
